webapp/dojo_news/src/App.spec.js

```diff
@@ -0,0 +1,221 @@
+import {shallowMount} from '@vue/test-utils';
+import App from "./App";
+import 'regenerator-runtime';
+import Vue from "vue";
+
+describe("News list", () => {
+    it('does not render any items on an initially empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+    });
+
+    it('renders items on an initially filled list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+
+        expect(wrapper.findAll(".newslistitem").length).toBe(3);
+    });
+    it('renders items correctly dependent on mutations of the list of NewsItems', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(1);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+
+        wrapper.vm.createNewsListItem("Test");
+        wrapper.vm.createNewsListItem("Test2");
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(2);
+    });
+});
+
+describe("List Empty Message", () => {
+    const listEmptyMessage = "Oh noes, the list is empty!";
+
+    it('displays a message on an initially empty list of NewsItems', () => {
+        const wrapperWithEmptyList = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapperWithEmptyList.text()).toContain(listEmptyMessage);
+    });
+
+    it('displays no message on an initially non-empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
```

Comment:
I like to use a setup method:
```js
const setupWrapper = (opts) => {
  const defaults = { /* ... */ }
  shallowMount(App, {...defaults, ...opts })
}
```

In you describe block:
```js
let propsData
describe('given no `items`', () => {
  beforeEach(() => {
    propsData = []
  })
})
```

In your test case:
```js
it('renders empty state', () => {
  wrapper = setupWrapper({ propsData })
})
```

Reply:
This way, you would add abstraction to the tests again. Isn't this kind of unwanted? Or better: Where do we want abstraction, where do we not?

Reply:
It certainly adds abstraction. There a folks out there who prefer an all-flat test hierarchy without any `describe`, `beforeEach` and friends. All code for test setup would have to be copied into every `test('... ', () => {})` case. This adds a lot of redundancy but has the benefit that the reader has to look into the implementation of the test case only to be able to understand the test.

I try to prevent false-negatives and I found an all flat-hierarchy can *also be error-prone*.

When I implement a feature, I do ["red, green, refactor"](https://www.codecademy.com/articles/tdd-red-green-refactor). In addition to the red test, I add one green test case which shares some test setup of the red test. This green test case helps me to know that the test setup is correct. If I had to maintain multiple test setups because I have an all-flat hierarchy, then I can make mistakes when I copy over the code.

A shared test setup for all test cases within a `describe` block helps to avoid this problem. I can always rely on 1 failing / 1 passing case when I change the setup.


Following this 1 green / 1 red pattern, I usually build up a tree-like hierarchy:
```
feature to be tested
  behaves like X
  given a condition A
    behaves like Y
    given condition B
      behaves like Z
    but given condition C
      behaves like X again
```

So I hope this explains why I'm in favour of hard-coded assertion values (low complexity) but also suggest something like a shared setup method (high complexity). In each `beforeEach` inside a `describe('condition A', () => {}), you would only implement changes for condition A.